### PR TITLE
Kernel: Simplify Storage subsystem concepts

### PR DESCRIPTION
**Pull request by supercomputer7**
This does the following 3 things:
1. Remove stale code such as the `ATAPIDiscDevice` class.
2. Allow each storage device to hold more than 16 partitions
3. Remove unnecessary `InterfaceType` enum-class.

This is a preparation before #14776 can be merged.

### CODE REVIEW

File: Kernel/Storage/StorageManagement.cpp [hunk 1]
```diff
@@ -241,8 +241,15 @@ MajorNumber StorageManagement::storage_type_major_number()
 }
 MinorNumber StorageManagement::generate_storage_minor_number()
 {
-    auto minor_number = s_device_minor_number.load();
-    s_device_minor_number++;
+    auto minor_number = s_storage_device_minor_number.load();
+    s_storage_device_minor_number++;
```
Comment (kleinesfilmroellchen) [inline Kernel/Storage/StorageManagement.cpp:245]: Even though this code is probably only called in one place on one kernel thread currently, it runs the risk of assigning the same minor number to two storage devices during parallel execution (e.g. device discovery at OS runtime). `fetch_add(1)` will definitely be safe for all circumstances.

File: Kernel/Storage/StorageManagement.cpp [hunk 2]
```diff
@@ -241,8 +241,15 @@ MajorNumber StorageManagement::storage_type_major_number()
...
+    return minor_number;
+}
+
+MinorNumber StorageManagement::generate_partition_minor_number()
+{
+    auto minor_number = s_partition_device_minor_number.load();
+    s_partition_device_minor_number++;
```
Comment (kleinesfilmroellchen) [inline Kernel/Storage/StorageManagement.cpp:252]: Same here, though I think this is less critical as within a device we'll probably always be initializing partitions in a simple loop.

File: Kernel/Storage/StorageManagement.cpp [hunk 3]
```diff
@@ -30,7 +30,8 @@
 namespace Kernel {
 
 static Singleton<StorageManagement> s_the;
-static Atomic<u32> s_device_minor_number;
+static Atomic<u32> s_storage_device_minor_number;
+static Atomic<u32> s_partition_device_minor_number;
```
Comment (kleinesfilmroellchen) [inline Kernel/Storage/StorageManagement.cpp:34]: If you're using the `MinorNumber` type alias in the APIs now, these can also just be `Atomic<MinorNumber>`, right?

**Review by kleinesfilmroellchen — CHANGES_REQUESTED** [review]
always yakbaiting me with refactors...

File: Kernel/Storage/StorageManagement.cpp (hunk 3 above)
Comment (supercomputer7) [reply Kernel/Storage/StorageManagement.cpp:34]: The compiler doesn't accept `MinorNumber` type as it is not "pure" primitive type like `u32`, so sadly we can't do this.

**Pull request merged**